azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -345,6 +354,133 @@ def test_connect_failure(self, mocker, stage, op, arbitrary_exception):
+        stage.pipeline_root.connected = True
+        connect_op.complete()
+
+    @pytest.mark.parametrize("connected", connected_and_disconnected)
+    @pytest.mark.it(
+        "Does not complete the operation if the op fails and the transport is no longer connected"
+    )
+    def test_failure_and_disconnect_no_complete(
+        self, mocker, stage, op, arbitrary_exception, connected
+    ):
+        stage.pipeline_root.connected = connected
+
+        stage.run_op(op)
+
+        # if we weren't previously connected, connect and reset mocks
+        if not connected:
+            self.complete_connection_and_reset_mock(stage)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] == op
+        stage.send_op_down.reset_mock()
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that the op didn't complete
+        assert not op.completed
+
+    @pytest.mark.parametrize("connected", connected_and_disconnected)
+    @pytest.mark.it(
+        "Sends down a new ConnectOperation if the op fails and the transport is no longer connected"
+    )
+    def test_failure_and_disconnect_new_connect_operation(
+        self, mocker, stage, op, arbitrary_exception, connected
+    ):
+        stage.pipeline_root.connected = connected
+
+        stage.run_op(op)
+
+        # if we weren't previously connected, connect and reset mocks
+        if not connected:
+            self.complete_connection_and_reset_mock(stage)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] == op
+        stage.send_op_down.reset_mock()
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that we got a new connect operation
+        assert stage.send_op_down.call_count == 1
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+
+    @pytest.mark.parametrize("connected", connected_and_disconnected)
+    @pytest.mark.it(
+        "Runs the operation a second time if the stage needs to reconnect after the op fails"
+    )
+    def test_failure_and_disconnect_resubmit_operation(
+        self, mocker, stage, op, arbitrary_exception, connected
+    ):
+        stage.pipeline_root.connected = connected
+
+        stage.run_op(op)
+
+        # if we weren't previously connected, connect and reset mocks
+        if not connected:
+            self.complete_connection_and_reset_mock(stage)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] == op
+        stage.send_op_down.reset_mock()
```

Comment:
>send_op_down [](start = 14, length = 12)

is it necessary to reset_mock instead of getting the call_args for each call and making sure the call_count is incrementing?